model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java

```diff
@@ -425,6 +425,10 @@ public UserSessionModel getUserSessionWithPredicate(RealmModel realm, String id,
 
         // Try lookup userSession from remoteCache
```

Comment:
I don't understand this method's logic. Line 421, `predicate.test(userSession)`, if the predicate returns `false`, why isn't the method returning `null`? Why does it continue to try to fetch the session from the cache? Are we expecting the predicate to return something different with the cached data?

Reply:
True, this is strange and it is worth another analysis. Please create a follow-up issue as it is IMHO unrelated to this issue.

Reply:
Issue created: https://github.com/keycloak/keycloak/issues/31115